app/src/main/java/com/bc/gordiansigner/service/WalletService.kt

```diff
@@ -34,34 +41,55 @@ class WalletService @Inject constructor(
         }.subscribeOn(Schedulers.computation())
     }
 
-    /** BEGIN -- simple mechanism getting & saving account by storing XPRVS **/
+    fun getHDKeyXprvs() =
+        fileStorageApi.SECURE.rxSingle { gateway ->
+            val privs = gateway.readOnFilesDir(XPRIV_KEY_FILE)
+            if (privs.isEmpty()) {
+                emptyList()
+            } else {
+                newGsonInstance().fromJson<List<String>>(String(privs))
+            }
+        }.map { set -> set.map { HDKey(it) } }
 
-    fun getLocalHDKeyXprvs() =
-        sharedPrefApi.SECURE.rxSingle { sharedPref ->
-            sharedPref.get(ROOT_XPRV_KEYS, Set::class)
-        }.map { set -> set.map { HDKey(it as String) } }
+    fun getHDKeyFingerprints() = fileStorageApi.SECURE.rxSingle { gateway ->
+        val fingerprints = gateway.readOnFilesDir(FINGERPRINT_FILE)
+        if (fingerprints.isEmpty()) {
+            emptyList()
+        } else {
+            newGsonInstance().fromJson<List<String>>(String(fingerprints))
+        }
+    }
 
     fun saveHDKeyXprv(hdKey: HDKey) =
-        getLocalHDKeyXprvs().map {
-            it.toMutableSet().apply {
-                add(hdKey)
+        getHDKeyXprvs().flatMap { keys ->
+            val keySet = keys.toMutableSet()
+            if (keySet.add(hdKey)) {
+                saveHDKeys(keySet).andThen(Single.just(hdKey))
+            } else {
+                Single.just(hdKey)
             }
-        }.flatMapCompletable { keys ->
-            saveHDKeyXprvs(keys)
-        }.andThen(Single.just(hdKey))
-
-    fun deleteKey(fingerprintHex: String) =
-        getLocalHDKeyXprvs().map {
-            it.filter { key -> key.fingerprintHex != fingerprintHex }.toSet()
-        }.flatMapCompletable { keys ->
-            saveHDKeyXprvs(keys)
         }
 
-    private fun saveHDKeyXprvs(keys: Set<HDKey>) =
-        sharedPrefApi.SECURE.rxCompletable { sharedPref ->
-            sharedPref.put(ROOT_XPRV_KEYS, keys.map { it.xprv }.toSet())
+    fun deleteHDKey(fingerprintHex: String) = getHDKeyXprvs().flatMapCompletable { keys ->
+        val fingerprints = keys.map { it.fingerprintHex }
+        if (fingerprints.contains(fingerprintHex)) {
+            val newKeys = keys.filterNot { it.fingerprintHex == fingerprintHex }.toSet()
+            saveHDKeys(newKeys)
+        } else {
+            Completable.complete()
         }
+    }
 
-    /** END - simple mechanism getting & saving account by storing XPRVS **/
+    private fun saveHDKeys(keys: Set<HDKey>) =
+        fileStorageApi.SECURE.rxCompletable { gateway ->
+            gateway.writeOnFilesDir(
+                XPRIV_KEY_FILE,
+                newGsonInstance().toJson(keys.map { it.xprv }).toByteArray()
+            )
+            gateway.writeOnFilesDir(
```

Comment:
I'm thinking of an idea to save fingerprint & xprv inside only a single file. I think it will be easier to manage and may avoid some mistakes in case of saving 1 of 2 file saving error or sometimes accidentally change one file but forget to change the other file.

Reply:
I intended to store xprivs in `fileStorageApi.SUPER_SECURE` so it has different security level as fingerprint. I will do it in next PR.